devices/src/virtio/vsock/epoll_handler.rs

```diff
@@ -53,15 +76,65 @@ where
     /// Walk the driver-provided RX queue buffers and attempt to fill them up with any data that we
     /// have pending.
     ///
-    fn process_rx(&mut self) {
-        error!("vsock: EpollHandler::process_rx() NYI");
+    fn process_rx(&mut self) -> bool {
+        debug!("vsock: epoll_handler::process_rx()");
+
+        let mut have_used = false;
+
+        while let Some(head) = self.rxvq.pop(&self.mem) {
+            let used_len = match VsockPacket::from_rx_virtq_head(&head) {
+                Ok(mut pkt) => {
+                    if self.backend.recv_pkt(&mut pkt).is_ok() {
+                        pkt.hdr().len() as u32 + pkt.len()
+                    } else {
+                        // We are using a consuming iterator over the virtio buffers, so, if we can't
+                        // fill in this buffer, we'll need to undo the last iterator step.
+                        self.rxvq.undo_pop();
+                        break;
+                    }
+                }
+                Err(e) => {
+                    warn!("vsock: RX queue error: {:?}", e);
+                    0
+                }
+            };
+
+            have_used = true;
+            self.rxvq.add_used(&self.mem, head.index, used_len);
+        }
+
+        have_used
     }
 
     /// Walk the dirver-provided TX queue buffers, package them up as vsock packets, and send them to
     /// the backend for processing.
     ///
-    fn process_tx(&mut self) {
-        error!("vsock: EpollHandler::process_tx() NYI");
+    fn process_tx(&mut self) -> bool {
+        debug!("vsock: epoll_handler::process_tx()");
+
+        let mut have_used = false;
+
+        while let Some(head) = self.txvq.pop(&self.mem) {
+            let pkt = match VsockPacket::from_tx_virtq_head(&head) {
+                Ok(pkt) => pkt,
+                Err(e) => {
+                    error!("vsock: error reading TX packet: {:?}", e);
+                    have_used = true;
+                    self.txvq.add_used(&self.mem, head.index, 0);
+                    continue;
+                }
+            };
+
+            if self.backend.send_pkt(&pkt).is_err() {
```

Comment:
Same with `recv_pkt`...isn t there the chance that we enter infinite behavior if we undo the pop?
Also, why not try to continue with the rest of the descriptors?
Is this behavior documented anywhere?